src/ui/public/agg_types/buckets/date_histogram.js

```diff
@@ -48,12 +50,22 @@ function getBounds(vis) {
 }
 
 function setBounds(agg, force) {
-  if (agg.buckets._alreadySet && !force) return;
-  agg.buckets._alreadySet = true;
-  const bounds = getBounds(agg.vis);
-  agg.buckets.setBounds(agg.fieldIsTimeField() && bounds);
+  const timeBuckets = getTimeBuckets(agg);
+  if (timeBuckets._alreadySet && !force) return;
+  timeBuckets._alreadySet = true;
+  timeBuckets.setBounds(agg.fieldIsTimeField() && getBounds(agg.vis));
 }
 
+function getTimeBuckets(aggConfig) {
+  if (timeBucketCache.has(aggConfig)) {
+    return timeBucketCache.get(aggConfig);
+  }
+  const buckets = new TimeBuckets();
+  timeBucketCache.set(aggConfig, buckets);
+  buckets.setInterval(getInterval(aggConfig));
+  setBounds(aggConfig);
```

Comment:
Does this need to be here? It seems like a side effect, and also makes the logic difficult to follow because `setBounds` and `getTimeBuckets` call each other. Is there a way to structure this so that the logic doesn't give the appearance of being circular?

Reply:
yup, this looks confusing :)

Reply:
I mean it seems to be circular, why the order matters. I tried not to touch the functioning of that code at all, since I found that it's a) very fragile and already caused some troubles in the past and b) hopefully at some point can be removed completely, since ES is already working on auto bucketing for histogram aggregations (https://github.com/elastic/elasticsearch/pull/28993).

If you think it's still worth until then to properly try to understand and make this code more beautiful I can give this a try.

Reply:
if you are afraid of doing this for 6.4 can you open a separate pr for 6.5 that will change this ?